src/Microsoft.OpenApi.Readers/ParsingContext.cs

```diff
@@ -1,30 +1,113 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT license. 
 
+using System;
 using System.Collections.Generic;
 using System.Linq;
 using Microsoft.OpenApi.Interfaces;
 using Microsoft.OpenApi.Models;
+using Microsoft.OpenApi.Readers.ParseNodes;
 using Microsoft.OpenApi.Readers.ReferenceServices;
+using SharpYaml.Serialization;
 
 namespace Microsoft.OpenApi.Readers
 {
     /// <summary>
-    /// Parsing context.
+    /// The Parsing Context holds temporary state needed whilst parsing an OpenAPI Document
     /// </summary>
     public class ParsingContext
     {
         private readonly Stack<string> _currentLocation = new Stack<string>();
+        private readonly Dictionary<string, IOpenApiReferenceable> _referenceStore = new Dictionary<string, IOpenApiReferenceable>();
+        private readonly Dictionary<string, object> _tempStorage = new Dictionary<string, object>();
+        private IOpenApiVersionService _versionService;
+        internal RootNode RootNode { get; set; }
+        internal List<OpenApiTag> Tags { get; private set; } = new List<OpenApiTag>();
+
+
+        /// <summary>
+        /// Initiates the parsing process.  Not thread safe and should only be called once on a parsing context
+        /// </summary>
+        /// <param name="yamlDocument"></param>
+        /// <param name="diagnostic"></param>
+        /// <returns>An OpenApiDocument populated based on the passed yamlDocument </returns>
+        internal OpenApiDocument Parse(YamlDocument yamlDocument, OpenApiDiagnostic diagnostic)
+        {
+            RootNode = new RootNode(this, diagnostic, yamlDocument);
```

Comment:
>new RootNode(this, diagnostic, yamlDocument [](start = 23, length = 43)

check the arguments?

Reply:
I'm going to defer doing this.  I'm not sure where we should do the argument check.  I'm thinking it would be better at the higher level OpenApiReader that is actually public.